src/environmentd/tests/sql.rs

```diff
@@ -903,100 +903,67 @@ fn test_tail_table_rw_timestamps() -> Result<(), Box<dyn Error>> {
-                        return false;
-                    }
-                }
-                // Only verify if row without data is last row
-                None => {
-                    if i + 1 != rows.len() {
-                        return false;
-                    }
-                }
-            }
-        }
-
-        if rows.len() != 2 {
-            return false;
-        }
-
-        // First row reflects write. Written rows have not progressed, and all
-        // writes occur at the same timestamp.
-        assert_eq!(rows[0].get::<_, Option<bool>>("mz_progressed"), Some(false));
-        // Two writes with the same data have their diffs compacted
-        assert_eq!(rows[0].get::<_, Option<i64>>("mz_diff"), Some(2));
-
-        // Second row reflects closing timestamp, manufactured by the read
-        assert_eq!(rows[1].get::<_, Option<bool>>("mz_progressed"), Some(true));
-        assert_eq!(rows[1].get::<_, Option<i64>>("mz_diff"), None);
-
-        true
-    };
-
     client_interactive.batch_execute("CREATE TABLE t1 (data text)")?;
 
     client_tail.batch_execute(
         "COMMIT; BEGIN;
          DECLARE c1 CURSOR FOR TAIL t1 WITH (PROGRESS);",
     )?;
 
-    // Keep trying until you either panic or are able to verify the expected behavior.
-    loop {
-        client_interactive.execute("BEGIN", &[])?;
-        client_interactive.execute("INSERT INTO t1 VALUES ($1)", &[&"first".to_owned()])?;
-        client_interactive.execute("INSERT INTO t1 VALUES ($1)", &[&"first".to_owned()])?;
-        client_interactive.execute("COMMIT", &[])?;
-        let _ = client_interactive.query("SELECT * FROM T1", &[])?;
-        client_interactive.execute("BEGIN", &[])?;
-        client_interactive.execute("INSERT INTO t1 VALUES ($1)", &[&"second".to_owned()])?;
-        client_interactive.execute("INSERT INTO t1 VALUES ($1)", &[&"second".to_owned()])?;
-        client_interactive.execute("COMMIT", &[])?;
-
-        let first_rows = client_tail.query("FETCH ALL c1", &[])?;
-        let first_rows_verified = verify_rw_pair(&first_rows, "first");
-
-        let _ = client_interactive.query("SELECT * FROM t1", &[])?;
-
-        let second_rows = client_tail.query("FETCH ALL c1", &[])?;
-        let second_rows_verified = verify_rw_pair(&second_rows, "second");
-
-        if first_rows_verified && second_rows_verified {
-            let first_write_ts = first_rows[0].get::<_, MzTimestamp>("mz_timestamp");
-            let first_closed_ts = first_rows[1].get::<_, MzTimestamp>("mz_timestamp");
-            assert!(first_write_ts < first_closed_ts);
-
-            let second_write_ts = second_rows[0].get::<_, MzTimestamp>("mz_timestamp");
-            let second_closed_ts = second_rows[1].get::<_, MzTimestamp>("mz_timestamp");
-            assert!(first_closed_ts <= second_write_ts);
-            assert!(second_write_ts < second_closed_ts);
-            break;
+    client_interactive.execute("BEGIN", &[]).unwrap();
```

Comment:
It was very unclear to me what this test was trying to test and the fact that everything was run in an infinite loop made me very suspect. It also seemed like some of the premises of this test no longer held, for example I'm not sure what the close write timestamp referred to. Maybe this was from when writing at a ts didn't also close it? Also, reads may not increase the timestamp but there are multiple background processes that can increase it in between reads.

I'm not sure why this started to fail with this PR, my best guess is that slowing down the reads made it so the timestamp would always be increased in-between reads. I tried to re-write this test to be deterministic and not failing.

Reply:
Yes, this test is pretty old and I never fully understood it either. It has been pretty finnicky anytime we change timestamp stuff.